cpp/src/fil/infer.cu

```diff
@@ -67,34 +67,50 @@ struct ArgMax {
   }
 };
 
+template <typename output_type, bool FULL, int NITEMS, typename tree_type>
+__device__ __forceinline__ vec<NITEMS, output_type> tree_leaf_output
+(tree_type tree, int n_rows, int (&leaves)[NITEMS]) {
+  vec<NITEMS, output_type> out(0);
+#pragma unroll
+  for (int j = 0; j < NITEMS; ++j) {
+    if (FULL || j < n_rows) {
+      /** dependent names are not considered templates by default,
+          unless it's a member of a current [template] instantiation.
+          alternatively, could have used .base_node::output<... */
+      out[j] = tree[leaves[j]].template output<output_type>();
```

Comment:
I'm having trouble parsing this line.... the intent to extract the leaf value seems clear enough, but could you explain a bit what `output<output_type>()` is doing?

Reply:
Unfortunately, it doesn't compile without `template`, see the comment above.

`output<output_type>()` just interprets the bits that store the leaf weight as `output_type` (currently either `float` or `int`) and returns it.